producer/writer/message_writer.go

```diff
@@ -242,6 +240,14 @@ func (w *messageWriterImpl) write(
 	return errFailAllConsumers
 }
 
+func randIndex(iterationIndexes []int, i int) int {
+	j := rand.Intn(i + 1)
+	// NB: we should only mutate the order in the iteration indexes and
+	// keep the order of consumer writers unchanged to prevent data race.
+	iterationIndexes[i], iterationIndexes[j] = iterationIndexes[j], iterationIndexes[i]
+	return j
```

Comment:
I think this should be `i` instead.

Reply:
Good catch, although it should be `iterationIndexes[i]` instead, added a test for this